backend/src/main/java/app/linguistai/bmvp/service/AccountService.java

```diff
@@ -157,4 +163,68 @@ private String encodePassword(String plainPassword) {
             throw e;
         }       
     }
+
+    public ResetToken generateEmailToken(String email) throws Exception {
+        try {
+            User user = accountRepository.findUserByEmail(email).orElse(null);
+            if (user == null) {
+                throw new NotFoundException("User with email [" + email + "] not found");
+            }
+            // invalidate previous reset tokens of user
+            List<ResetToken> resetTokens = resetTokenRepository.findAllByUser(user);
+            for (ResetToken resetToken: resetTokens) {
+                resetToken.setUsed(true);
+            }
+            // create a new reset token
+            ResetToken resetToken = new ResetToken(user);
+            return resetTokenRepository.save(resetToken);
+        } catch (Exception e) {
+            System.out.println("Email token generation exception for email");
+            throw e;
+        }
+    }
+
+    public boolean validateResetCode(String email, String resetCode, boolean invalidate) throws Exception{
+        try {
+            User user = accountRepository.findUserByEmail(email).orElse(null);
+            if (user == null) {
+                throw new NotFoundException("User with email [" + email + "] not found");
+            }
+
+            ResetToken resetToken = resetTokenRepository.findByUserAndResetCode(user, resetCode).orElse(null);
+            if(resetToken == null){
```

Comment:
Can you put spaces between "if(resetToken == null){" -> "if (resetToken == null) {"

Reply:
Reformatting the code will be done in a different task - along with using tabs for indentation.

Reply:
Reformatting the code will be done in a different task - along with using tabs for indentation.